---
title: "Individual_datasets_integration"
output: html_document
date: "2024-07-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Sample 4 : Colonic Stroma. GSE172261

DSS+WT

Time: 7 day cycles * 3 = 21 days

Mice were subjected to 3 repetitive cycles of DSS displayed progressive accumulation of immune cell infiltrates associated with excessive deposition of collagen fibers. Lamina propria cells from water-fed and DSS-fed mice were isolated using enzymatic digestion, and enriched for stromal cells by FACS using antibodies excluding hematopoietic cells (CD45), epithelial cells (EpCAM), and erythrocytes. Prepared single cell suspensions were then profiled using the 10x Chromium V2 droplet-based single cell RNA sequencing platform. Tissues were collected from proximal and distal colon.


```{r}
library(Seurat)
library(readr)

Colonic_Stroma_DSS <- ReadMtx(
  mtx = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_matrix.mtx", features = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_genes.tsv",
  cells = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_barcodes.tsv"
)

Colonic_Stroma_DSS <- CreateSeuratObject(counts = Colonic_Stroma_DSS, min.cells = 3, min.features = 200)

# Reading metadat
metadata<-readr::read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_metadata.tsv")
print(metadata)

library(dplyr)
library(tibble)
# Set row names of metadata to cell_barcode
metadata <- metadata %>%
  tibble::column_to_rownames(var = "cell_barcode")

# Ensure the barcodes in metadata match the Seurat object
common_barcodes <- intersect(rownames(metadata), colnames(Colonic_Stroma_DSS))

# Subset metadata to only include common barcodes
metadata <- metadata[common_barcodes, ]

# Add metadata to Seurat object
Colonic_Stroma_DSS <- AddMetaData(Colonic_Stroma_DSS, metadata)

#Quality Control
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Colonic_Stroma_DSS[["percent.mt"]] <- PercentageFeatureSet(Colonic_Stroma_DSS, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Colonic_Stroma_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Colonic_Stroma_DSS, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Colonic_Stroma_DSS, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Colonic_Stroma_DSS@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Colonic_Stroma_DSS <- PercentageFeatureSet(Colonic_Stroma_DSS, "^Rp", col.name = "percent_ribo")
VlnPlot(Colonic_Stroma_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Colonic_Stroma_DSS <- subset(Colonic_Stroma_DSS, subset = nFeature_RNA > 300 & nFeature_RNA < 3400 & percent.mt < 8 & percent_ribo <25)

View(Colonic_Stroma_DSS@meta.data)

#Create one column with all the cell types
library(dplyr)

Colonic_Stroma_DSS$celltype <- Colonic_Stroma_DSS@meta.data %>% 
  mutate(celltype = coalesce(Celltype_Stroma_Atlas,Celltype_Endothelium_Atlas,Celltype_Fibroblast_Atlas))

Colonic_Stroma_DSS$Age <- "12 weeks"

obj <- Colonic_Stroma_DSS

obj[["RNA"]] <- split(obj[["RNA"]], f = obj$batch)
obj

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("batch", "celltype"))

obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj, resolution = 0.5, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")
p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("batch", "celltype", "harmony_clusters", "Treatment"),
  combine = FALSE, label.size = 2, label = TRUE
)
library(patchwork)
 wrap_plots(c(p1), ncol = 2, nrow = 2)
 
 DimPlot(obj, reduction = "umap.harmony", group.by = "Site", combine = FALSE, label.size = 2, label = TRUE
)

obj_Colonic_Stroma_DSS <- obj

# Replace "H2O" with "ctrl" in the Treatment metadata of the Seurat object
obj_Colonic_Stroma_DSS$Treatment <- ifelse(obj_Colonic_Stroma_DSS$Treatment == "H2O", "ctrl", obj_Colonic_Stroma_DSS$Treatment)

obj_Colonic_Stroma_DSS$cellType <- obj_Colonic_Stroma_DSS$celltype

# Create a new column 'celltype_general' with generalized cell types
obj_Colonic_Stroma_DSS@meta.data <- obj_Colonic_Stroma_DSS@meta.data %>%
  mutate(celltype_general = case_when(
    cellType %in% c("Fibroblast 1a", "Fibroblast 1b", "Fibroblast 1c") ~ "Fibroblast 1",
    cellType %in% c("SMC 1", "SMC 2") ~ "SMC",
    cellType %in% c("Fibroblast 2c", "Fibroblast 2a", "Fibroblast 2b") ~ "Fibroblast 2",
    cellType %in% c("Fibroblast 3a", "Fibroblast 3b") ~ "Fibroblast 3",
    cellType %in% c("BEC 1", "BEC 2") ~ "BEC",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))

# Create a new column 'celltype_general' with generalized cell types
library(dplyr)
obj_Colonic_Stroma_DSS@meta.data <- obj_Colonic_Stroma_DSS@meta.data %>%
  mutate(celltype_integration = case_when(
    cellType %in% c("Fibroblast 1a", "Fibroblast 1b", "Fibroblast 1c", "Fibroblast 2c", "Fibroblast 2a", "Fibroblast 2b", "Fibroblast 1","Fibroblast 3a", "Fibroblast 3b") ~ "Fibroblast",
    cellType %in% c("SMC 1", "SMC 2") ~ "SMC",
    cellType %in% c("BEC 1", "BEC 2") ~ "BEC",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))

#Fibroblast 1 is CD81 pos
#Fibroblast 2 is CD81 pos ->PDGFRalpha_lo_CD81_neg_Fibroblasts
#Fibroblast 3 is CD81 pos ->PDGFRalpha_hi_Fibroblasts

#How was the annotation of the dataset given?
#https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3001532 

saveRDS(obj_Colonic_Stroma_DSS, file = "Colonic_Stroma.rds")

```


What do each of the fibroblast groups correspond to?
```{r}
# Remember to switch to raw data for DEG
DefaultAssay(Colonic_Stroma) <- "RNA"
# Find DEGs in celseq
#celseq.markers <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 1",group.by = "tech", logfc.threshold = 0.25, only.pos = TRUE)

# Find DEGs in alpha
Idents(Colonic_Stroma) <- "celltype_general"
markers.fb1 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 1", logfc.threshold = 0.25, only.pos = TRUE)
markers.fb2 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 2", logfc.threshold = 0.25, only.pos = TRUE)
markers.fb3 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 3", logfc.threshold = 0.25, only.pos = TRUE)


# Find conserved DEGs among techs
all.markers <- FindConservedMarkers(markers.fb1, ident.1 = c("Fibroblast 1"), grouping.var = "Treatment")
```
```{r}
# Find DEGs for all techs
Idents(pancreas.integrated) <- "tech"
all.markers <- FindAllMarkers(pancreas.integrated, logfc.threshold = 0.25, only.pos =
TRUE)
top_10_marker <- all.markers %>% group_by(cluster) %>% top_n(n =10, avg_log2FC)
head(top_10_marker)
# Draw heatmap
DoHeatmap(pancreas.integrated,features = top_10_marker$gene,slot="counts", size = 4) + scale_fill_gradientn(colors = RColorBrewer::brewer.pal(n = 9, name = "RdBu"))
```

```{r}
) <- "tech" VlnPlot(pancreas.integrated, features =c("REG1A"))
VlnPlot(pancreas.integrated, features =c("REG1A"), group.by = "celltype") VlnPlot(pancreas.integrated, features = c("REG1A"),split.by = "tech")
pancreas.integrated.sub <- subset(pancreas.integrated, idents=c("celseq","celseq2")) VlnPlot(pancreas.integrated.sub, features = c("REG1A"),group.by = "celltype",
split.by = "tech", cols=c("red","grey","blue"), pt.size = 0) FeaturePlot(pancreas.integrated, features = c("REG1A"), split.by = "tech",
max.cutoff = 3, cols = c("grey", "red"))
```




Sample 6. Epithelial LP
We first compared the epithelial healing effects of RSPO2 and a Wnt mimetic with broad Fzd-specificity in an acute Dextran Sodium Sulfate (DSS) mouse colitis model. Guided by Fzd expression patterns in the colon epithelium, we also examined the effects of Wnt mimetics with sub-family Fzd-specificities. 
 
Read in the matrices for the wild type samples

```{r}
count_table <- read.table("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Epithelial_LP/GSE201723_SRZ_colon_mat.tsv")
Epithelial.LP <- CreateSeuratObject(counts = count_table, min.cells = 3, min.features = 300)

metadata <- read.table("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Epithelial_LP/GSE201723_SRZ_colon_metaData.tsv")
head(metadata)
Epithelial.LP <- AddMetaData(Epithelial.LP, metadata)

```

QC
```{r}
#Quality Control
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Epithelial.LP[["percent.mt"]] <- PercentageFeatureSet(Epithelial.LP, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Epithelial.LP, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Epithelial.LP, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Epithelial.LP, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Epithelial.LP@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Epithelial.LP <- PercentageFeatureSet(Epithelial.LP, "^Rp", col.name = "percent_ribo")
VlnPlot(Epithelial.LP, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Epithelial.LP <- subset(Epithelial.LP, subset = nFeature_RNA > 300 & nFeature_RNA < 5000 & percent.mt < 4 & percent_ribo <10)

View(Epithelial.LP@meta.data)

```

Keep only uninjured and DSS without treatment. "We first assessed the effect of DSS injury by comparing the DSS, anti-GFP condition with the uninjured condition" 
```{r}
# Subset the Seurat object to keep only cells corresponding to 'Uninj' or 'antiGFP'
Epithelial.LP_filtered <- subset(
  Epithelial.LP, 
  subset = batch %in% c("d5_24h_Uninj_1", "d5_24h_Uninj_2", "d6_48h_Uninj_1", "d6_48h_Uninj_2", "d5_24h_aGFP_1", "d5_24h_aGFP_2", "d5_24h_aGFP_3", "d6_48h_aGFP_1", "d6_48h_aGFP_2", "d6_48h_aGFP_3")
)

Epithelial.LP_filtered$condition <- NA

# Assign values based on the 'batch' column
Epithelial.LP_filtered$condition[grep("Uninj", Epithelial.LP_filtered$batch)] <- "Uninjured"
Epithelial.LP_filtered$condition[grep("aGFP", Epithelial.LP_filtered$batch)] <- "antiGFP"

# Optionally, you can check the unique values of the new 'condition' column
unique(Epithelial.LP_filtered$condition)

```

```{r}

obj <- Epithelial.LP_filtered

obj[["RNA"]] <- split(obj[["RNA"]], f = obj$condition)
obj

obj <- SCTransform(obj, vars.to.regress = c("percent.mt", "percent_ribo"), verbose = FALSE)
obj <- RunPCA(obj)

obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 1, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("batch", "cellType"))

Layers(obj)

obj <- Seurat::IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj, resolution = 0.5, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")

# Replace "antiGFP" with "DSS" in the condition metadata
obj$condition <- ifelse(obj$condition == "antiGFP", "DSS", obj$condition)

p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("batch", "cellType", "harmony_clusters", "condition"),
  combine = FALSE, label.size = 2, label = TRUE
)
 wrap_plots(c(p1), ncol = 2, nrow = 2)
 
 DimPlot(obj, reduction = "umap.harmony", group.by = "tissueLayer", combine = FALSE, label.size = 2, label = TRUE
)
 
DimPlot(obj, reduction = "umap.harmony", group.by = "cellType", combine = FALSE, label.size = 2, label = TRUE
)

obj.Epithelial.LP_filtered <- obj

obj.Epithelial.LP_filtered$Treatment <- obj.Epithelial.LP_filtered$condition

# Replace "H2O" with "ctrl" in the Treatment metadata of the Seurat object
obj.Epithelial.LP_filtered$Treatment <- ifelse(obj.Epithelial.LP_filtered$Treatment == "Uninjured", "ctrl", obj.Epithelial.LP_filtered$Treatment)

 # Create a new column 'celltype_integration' with generalized cell types
library(dplyr)
obj.Epithelial.LP_filtered@meta.data <- obj.Epithelial.LP_filtered@meta.data%>%
mutate(celltype_integration = case_when(
cellType %in% c("Bcell1_IgM", "Bcell2_IgM") ~ "Bcell_IgM",
cellType %in% c("Dendritic", "ActDendri") ~ "DC",
cellType %in% c("InjuryMono1", "InjuryMono2") ~ "InjuryMono",
cellType %in% c("CryptFB1", "CryptFB2", "BaseCryptFB1", "InjuryCryptFB1", "InjuryCryptFB2", "BaseCryptFB2",   "FB3" ) ~ "Fibroblast",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("AltEnteroPC", "EnteroPrecur", "AltEnterocyte", "ImmEntero1", "ImmEntero2", "Enterocyte", "AltEnterocyte2",  "AltEnterocyte3") ~ "Enterocyte",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("TA1", "TA2" ) ~ "TA",
cellType %in% c("Glia1", "Glia2" ) ~ "Glia",
TRUE ~ cellType  # Keeps the original cellType for those not specified
))

saveRDS(obj.Epithelial.LP_filtered, file = "Epi_Imm_Stro_colon.rds")

```



```{r}

 # Epithelial.LP.ctrl <- subset(
  #Epithelial.LP, 
  #subset = batch %in% c("d5_24h_Uninj_1", "d5_24h_Uninj_2", "d6_48h_Uninj_1", "d6_48h_Uninj_2"))


# run sctransform
#Epithelial.LP.ctrl <- SCTransform(Epithelial.LP.ctrl, vars.to.regress = c("percent.mt", "percent_ribo"), verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
#Epithelial.LP.ctrl <- RunPCA(Epithelial.LP.ctrl, verbose = FALSE)
#Epithelial.LP.ctrl <- RunUMAP(Epithelial.LP.ctrl, dims = 1:30, verbose = FALSE)

#Epithelial.LP.ctrl <- FindNeighbors(Epithelial.LP.ctrl, dims = 1:30, verbose = FALSE)
#Epithelial.LP.ctrl <- FindClusters(Epithelial.LP.ctrl, verbose = FALSE, resolution = 0.5)
#DimPlot(Epithelial.LP.ctrl, label = TRUE, group.by = "cellType")

```

To convert to Python
```{r}
library(Seurat)

# Path to the main directory
main_dir <- "/Users/andson/Desktop/Mouse Atlas Datasets/Colon/"


list.files(main_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx

library(Seurat)
sample1 <- Read10X(data.dir = main_dir)

```

```{r}
sample1 <- Read10X(data.dir = "/Users/andson/Desktop/Mouse Atlas Datasets/Small Instestine/GSE218311/GSE218311_RAW/GSM6739150_2_p28wt_filtered_feature_bc_matrix")
sample1 <-  CreateSeuratObject(counts = sample1, project = "GSM6739150" ) 
 
sample2 <- Read10X(data.dir = "/Users/andson/Desktop/Mouse Atlas Datasets/Small Instestine/GSE218311/GSE218311_RAW/GSM6739151_3_p28wt_filtered_feature_bc_matrix")
sample2 <-  CreateSeuratObject(counts = sample2, project = "GSM6739151")

sample3 <- Read10X(data.dir = "/Users/andson/Desktop/Mouse Atlas Datasets/Small Instestine/GSE218311/GSE218311_RAW/Sample3")
sample3 <-  CreateSeuratObject(counts = sample3, project = "GSM6739149")

sample4 <- Read10X(data.dir = "/Users/andson/Desktop/Mouse Atlas Datasets/Small Instestine/GSE218311/GSE218311_RAW/GSM6739152_4_p28wt_filtered_feature_bc_matrix")
sample4 <-  CreateSeuratObject(counts = sample4, project = "GSM6739152")
```

Merge Objects
```{r}
CD4T.SI <- merge(sample1, y = c(sample2, sample3, sample4), add.cell.ids = c("GSM6739150", "GSM6739151","GSM6739149", "GSM6739152"), project = "CD4+ T cell")
CD4T.SI
```

```{r}
obj_Epithelial.LP_filtered<- obj.Epithelial.LP_filtered

obj_Epithelial.LP_filtered$dataset<- "Dataset_1"
Stromal_ctrl_DSS$dataset <- "Dataset_2" 
obj_Colonic_Stroma_DSS$dataset<- "Dataset_3"


# Load the Seurat library
library(Seurat)

# Combine 'dataset' and 'Treatment' columns to create the 'COVARIATE' column
obj_Epithelial.LP_filtered$covariate <- paste(obj_Epithelial.LP_filtered$dataset, obj_Epithelial.LP_filtered$Treatment, sep = "_")


Stromal_ctrl_DSS$covariate <- paste(Stromal_ctrl_DSS$dataset, Stromal_ctrl_DSS$Treatment, sep = "_")


obj_Colonic_Stroma_DSS$covariate <- paste(obj_Colonic_Stroma_DSS$dataset, obj_Colonic_Stroma_DSS$Treatment, sep = "_")


seu <- merge(obj_Epithelial.LP_filtered, y = c(Stromal_ctrl_DSS, obj_Colonic_Stroma_DSS), add.cell.ids = c("D1", "D2", "D3"), project = "colon_atlas")

sceasy::convertFormat(seu, from="seurat", to="anndata",
                       outFile='seu.h5ad')


library(sceasy)
sceasy::convertFormat(Stromal_ctrl_DSS, from="seurat", to="anndata",
                       outFile='Stromal_ctrl_DSS.h5ad')
sceasy::convertFormat(obj_Colonic_Stroma_DSS, from="seurat", to="anndata",
                       outFile='Stromal_ctrl_DSS.h5ad')
sceasy::convertFormat(obj_Epithelial.LP_filtered, from="seurat", to="anndata",
                       outFile='Stromal_ctrl_DSS.h5ad')


```


###SMART-SEQ Data

#Colonic Cells. GSE148794

```{r}
library(readr)

df <- read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_cells/GSE148794_tc_ibd.count_table.tsv")

# Move the first column (named `...1`) to row names
df <- df %>%
  column_to_rownames(var = "...1")

metadata <-read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_cells/GSE148794_tc_ibd.metadata.tsv")

```

```{r}
# Split the 'Title' column into 'Treatment' and 'Sample' columns
metadata <- metadata %>%
  separate(Title, into = c("Prefix", "Treatment", "Sample_Suffix"), sep = "\\.", remove = FALSE) %>%
  separate(Sample_Suffix, into = c("Sample", "Rest"), sep = "_", remove = FALSE) %>%
  select(-Prefix, -Rest)


```


Sample 3. IEC_Stroma_RAW GSE163638

Mouse colitis was induced by administration of 2% w/v of dextran sodium sulfate (DSS, TdB Consultancy, MW = 40 kDa) dissolved in drinking water ad libitum for 7 days, followed by 7 days of regular water. General health and body weight were monitored regularly. On day 14, colons from untreated and DSS-treated mice were harvested, measured, and processed for spatial transcriptomic studies.

DSS only

To analyze the impact B cells might have on intestinal epithelial cells and stromal cells during recovery after intestinal injury the transcriptional profile of these mice was analysed in mice depleted of B cells and control mice on day 14 after DSS colitis.

Just using ctrl sample. Not cell b depleted
```{r}
Stromal_ctrl_DSS <- ReadMtx(
  mtx = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-matrix.mtx", features = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-features.tsv",
  cells = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-barcodes.tsv"
)

Stromal_ctrl_DSS <- CreateSeuratObject(counts = Stromal_ctrl_DSS, min.cells = 3, min.features = 200)

#Quality Control
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Stromal_ctrl_DSS[["percent.mt"]] <- PercentageFeatureSet(Stromal_ctrl_DSS, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Stromal_ctrl_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Stromal_ctrl_DSS, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Stromal_ctrl_DSS, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Stromal_ctrl_DSS@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Stromal_ctrl_DSS <- PercentageFeatureSet(Stromal_ctrl_DSS, "^Rp", col.name = "percent_ribo")
VlnPlot(Stromal_ctrl_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Stromal_ctrl_DSS <- subset(Stromal_ctrl_DSS, subset = nFeature_RNA > 300 & nFeature_RNA < 4000 & percent.mt < 60 & percent_ribo <55)

View(Stromal_ctrl_DSS@meta.data)


# run sctransform
Stromal_ctrl_DSS <- SCTransform(Stromal_ctrl_DSS, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal_ctrl_DSS <- RunPCA(Stromal_ctrl_DSS, verbose = FALSE)
Stromal_ctrl_DSS <- RunUMAP(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)

Stromal_ctrl_DSS <- FindNeighbors(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)
Stromal_ctrl_DSS <- FindClusters(Stromal_ctrl_DSS, verbose = FALSE, resolution = 0.5)
DimPlot(Stromal_ctrl_DSS, label = TRUE)
Stromal_ctrl_DSS$Treatment <- "DSS"
```
VERY ELEVATED MITHOCONDRIALCONTAMINATION. Include?

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
stromal.markers <- FindAllMarkers(Stromal_ctrl_DSS, only.pos = TRUE)
stromal.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Stromal_ctrl_DSS$Treatment <- "DSS"
```
```{r}
stromal.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(Stromal_ctrl_DSS, features = top10$gene) + NoLegend()

Stromal_ctrl_DSS$Treatment <- "DSS"
```
From paper
```{r}
Bcell <- c("Ms4a1", "Cd19", "Faim3", "Cd79b")

MNPs <- c("Lyz2", "Aif1", "Csf1r")

Enteroendocrine <- c("Chga", "Chgb", "Resp18", "Tph1", "Snap25", "Pcsk1n", "Sct", "Ddc", "Slc18a1", "Syt7", "Krt7", "Ptprn2", "Neurod1", "Insm1", "Krt20", "Rundc3a", "Ptprn", "Wif1", "Gsdma", "Scg5", "Rab3c", "Cadps", "Cacna1a", "Lmx1a", "Fam183b")

Endothelial <- c("Flt1", "Plvap", "Fabp4", "Ly6c1", "Ptprb", "Eltd1", "Egfl7", "Podxl", "Gpr116", "Kdr", "Pecam1", "Esam", "Emcn", "Cd36", "Ace", "Tmem88", "Cd93", "Jam2", "Cd34", "Clec14a", "Igfbp7", "Ly6a", "Cdh5", "Ehd4", "Sema7a")

Lymphatic <- c("Mmrn1", "Lyve1", "Reln", "Flt4", "Cldn5", "Gpr97", "Fgl2", "Stab1", "Cp", "Lbp", "Ecscr", "Timp3", "Slc45a3", "Ackr2", "Mmrn2", "Sema3d", "Sdpr", "Pard6g", "Fxyd6", "Ece1", "Hsd3b7", "Tbx1", "Thy1", "Galnt18", "Prelp")

Stromal <- c("Pdgfra", "Dpt", "Fbln1", "Dcn", "Lum", "Col1a2", "Col1a1", "Col3a1", "Gsn", "Mmp2", "Htra3", "Spon2", "Efemp1", "Adamdec1", "Clec3b", "Svep1", "Col6a2", "Gpx3", "Mgp", "Col6a1", "Pcolce", "Bgn", "Cygb", "Bicc1", "Col5a1")

Tcell <- c("Trbc2", "Cd3g", "Cd3d", "Trbc1", "Nkg7", "Ms4a4b", "Lck", "Skap1", "Ctsw", "Cd3e", "Zap70", "Xcl1", "Lat", "Ccl5", "Gimap4", "Il2rb", "AW112010", "Cd96", "Cd160", "Cxcr6", "Ptprcap", "Cd7", "Itk", "Klrd1", "Cd247")

Myofibroblast2 <- c("Rgs4", "Notch3", "Des", "Ndufa4l2", "Rgs5", "Pdgfrb", "Pln", "Rasl12", "Nrip2", "Mfge8", "Olfr558", "Sparcl1", "Myl9", "Itga7", "Acta2", "Gm13889", "Tagln", "Mustn1", "Tpm2", "Kcnab1", "Myh11", "Tpm1", "Cald1", "Gucy1b3", "Zak")

AbsAndSecCell <- c("Lypd8", "Dmbt1", "Prss32", "Tspan1", "Phgr1", "Lgals4", "Ckmt1", "Ceacam1", "2610528A11Rik", "Pigr", "Muc3", "Tspan8", "Oit1", "Muc13", "Elf3", "Guca2a", "Ppp1r1b", "Fabp2", "Cldn7", "Gpx2", "Krt8", "Cldn3", "Tm4sf20", "Tmem45b", "Myo1a")

PlasmacytoidDC <- c("Siglech", "Dntt", "Ccr9", "Cd300c", "Irf8", "Mpeg1", "Cox6a2", "Pld4", "Bst2", "Lair1", "Rnase6", "Runx2", "Kmo", "Ctsh", "2810442I21Rik", "Ly6c2", "Grn", "Spns3", "Flt3", "Snx5", "Smim5", "Bcl11a", "Clec10a", "Ly6d", "Ppfia4")

Granulocyte <- c("S100a9", "S100a8", "Hdc", "Retnlg", "Irg1", "G0s2", "Cxcl2", "1100001G20Rik", "Clec4e", "Clec4d", "Trem1", "Il1b", "Cd14", "Il1r2", "Slfn4", "Cxcr2", "Wfdc17", "Mxd1", "Mmp9", "Gp49a", "Csf3r", "1810033B17Rik", "Asprv1", "Lilrb4", "Hp")

PlasmaCell <- c("Igj", "Igha", "Igkc", "Iglv1", "Eaf2", "Derl3", "Ccr10", "Mzb1", "Tnfrsf17", "Iglc2", "Fam46c", "Trp53inp1", "Txndc5", "Pou2af1", "Edem1", "Glipr1", "Sec11c", "Xbp1", "Cacna1s", "Creld2", "Prg2", "Fkbp11", "Igkj5", "Iglv3", "Ighj4")

InterstitialCellOfCajal <- c("Pcp4", "Ano1", "Kit", "Kcnd3", "Grem2", "Foxp2", "Nog", "Cckar", "Ctxn3", "Lrig1", "Gja1", "Ak5", "Dpp10", "Cdh11", "Enpep", "Hand1", "Cpe", "Fgfr2", "Col23a1", "Slc4a4", "Clmp", "Cacna1d", "Sema3c", "Ano2", "Kit")

Myofibroblast1 <- c("Hhip", "Actg2", "Cnn1", "Postn", "Myh11", "Tpm2", "Mfap5", "Acta2", "Rbp4", "Mylk", "Pitx1", "Tagln", "Fhl1", "Cpxm2", "Pdlim3", "Lpp", "Flna", "Myl9", "Csrp1", "Tpm1", "Ppp1r14a", "F2r", "Tgfbr3", "Pgm5", "Tcf21")

GlialCell <- c("Plp1", "Scn7a", "Kcna1", "Lgi4", "Ank3", "Prnp", "Chl1", "Matn2", "Sostdc1", "Gpr37l1", "Kcna6", "Sema3b", "Cadm4", "Abca8a", "Sox10", "Art3", "S100b", "Abca8b", "Slc35f1", "Apod", "Cdh19", "Aspa", "L1cam", "Sfrp1", "Entpd2")

```


```{r}
# Feature plot - visualize feature expression in low-dimensional space
Idents(Stromal_ctrl_DSS) <- "SCT_snn_res.0.5"

FeaturePlot(Stromal_ctrl_DSS, features = c("Epcam", "Pdgfra", "Ptprc"))

FeaturePlot(Stromal_ctrl_DSS, features = "Pdgfra")
FeaturePlot(Stromal_ctrl_DSS, features = "Ptprc")


 Stromal_ctrl_DSS$Treatment <- "DSS"
```

There shouldn't be immune cells as only stromal and epithelial were gated. Remove cluster 13/clean cells that express ptprc
```{r}
FeaturePlot(Stromal_ctrl_DSS, features = "Ptprc")
#cluster 13
```

```{r}

```

```{r}
# run sctransform
Stromal_ctrl_DSS <- SCTransform(Stromal_ctrl_DSS, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal_ctrl_DSS <- RunPCA(Stromal_ctrl_DSS, verbose = FALSE)
Stromal_ctrl_DSS <- RunUMAP(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)

Stromal_ctrl_DSS <- FindNeighbors(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)
Stromal_ctrl_DSS <- FindClusters(Stromal_ctrl_DSS, verbose = FALSE, resolution = 0.5)
DimPlot(Stromal_ctrl_DSS, label = TRUE)
```
Epithelial.sub <- subset(Stromal_ctrl_DSS, idents = c(0, 2, 12, 7, 8, 9, 10))

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
stromal.markers <- FindAllMarkers(Stromal_ctrl_DSS, only.pos = TRUE)
stromal.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r}
stromal.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(Stromal_ctrl_DSS, features = top10$gene) + NoLegend()
```

```{r}
FeaturePlot(Stromal_ctrl_DSS, features = c("Epcam", "Pdgfra", "Ptprc"))

FeaturePlot(Stromal_ctrl_DSS, features = "Pdgfra")
FeaturePlot(Stromal_ctrl_DSS, features = "Epcam")
```
```{r}
#Subset for easier annotation
Epithelial.sub <- subset(Stromal_ctrl_DSS, idents = c(0, 2, 12, 7, 8, 9, 10))
Stromal.sub <- subset(Stromal_ctrl_DSS, idents = c(0, 2, 12, 7, 8, 9, 10), invert = TRUE)
```


```{r}
# Stromal
Stromal.sub <- SCTransform(Stromal.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal.sub <- RunPCA(Stromal.sub, verbose = FALSE)
Stromal.sub <- RunUMAP(Stromal.sub, dims = 1:30, verbose = FALSE)

Stromal.sub <- FindNeighbors(Stromal.sub, dims = 1:30, verbose = FALSE)
Stromal.sub <- FindClusters(Stromal.sub, verbose = FALSE, resolution = 0.4)
DimPlot(Stromal.sub, label = TRUE)
```
```{r}
Stromal.sub.marker <- FindAllMarkers(Stromal.sub, only.pos = TRUE)
Stromal.sub.marker %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Stromal.sub.marker %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(Stromal.sub, features = top10$gene) + NoLegend()
```


```{r}
# Epithelial
Epithelial.sub <- SCTransform(Epithelial.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Epithelial.sub <- RunPCA(Epithelial.sub, verbose = FALSE)
Epithelial.sub <- RunUMAP(Epithelial.sub, dims = 1:30, verbose = FALSE)

Epithelial.sub <- FindNeighbors(Epithelial.sub, dims = 1:30, verbose = FALSE)
Epithelial.sub <- FindClusters(Epithelial.sub, verbose = FALSE, resolution = 0.5)
DimPlot(Epithelial.sub, label = TRUE)

```

```{r}
Epithelial.sub.markers <- FindAllMarkers(Epithelial.sub, only.pos = TRUE)
Epithelial.sub.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Epithelial.sub.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10

DoHeatmap(Epithelial.sub, features = top10$gene) + NoLegend()
```


#Annotate the whole dataset:
```{r}
#Reference
unique(Epi_Imm_Stro_colon$celltype_integration)
#Subset to keep only relevant cells
ref <- Epi_Imm_Stro_colon

ref <- ref[,grepl('Fibroblast|MyoFB|MuralLikeFB|SmoothMuscle|VascSmMuscle1|VascSmMuscle2|ICCajal|Pericytes|VascEndo|LymphEndo|StemCell|Enterocyte|Goblet|Tuft|EnteroEnd1|EnteroEnd2', ref$celltype_integration)]

ref<- as.SingleCellExperiment(ref, assay = "SCT")

raw_counts <- LayerData(Stromal_ctrl_DSS, assay = "SCT") 

library(scran)
library(SingleR)
# 3. Run SingleR
#With only DSS subset of reference
ct_ann <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
```

```{r}
Stromal_ctrl_DSS$celltype_integration <- ct_ann$pruned.labels[match(rownames(Stromal_ctrl_DSS@meta.data), rownames(ct_ann))]
DimPlot(Stromal_ctrl_DSS, reduction = 'umap', group.by = 'celltype_integration')

```

```{r}
plotScoreHeatmap(ct_ann)
# Inspect quality of the predictions
plotDeltaDistribution(ct_ann, ncol = 4, dots.on.top = T)
```


```{r}
saveRDS(Stromal_ctrl_DSS, "Stromal_DSS.rds")
```

#Annorate Stromal cells
reference: obj_Colonic_Stroma_DSS

As the dataset to be annotated contains only DSS stimulated samples, subset the reference
```{r}
#ref.sub <- subset(obj_Colonic_Stroma_DSS, Treatment == "DSS")
#ref.sub<- JoinLayers(ref.sub)
ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(ref.sub)
unique(ref.sub$celltype_integration)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")
```


```{r}
#ref<- obj_Colonic_Stroma_DSS
#ref<- JoinLayers(ref.sub)

#Colonic_Stroma_DSS<- SCTransform(Colonic_Stroma_DSS, vars.to.regress = "percent.mt")
#unique(ref$celltype)
#ref<- as.SingleCellExperiment(ref, assay = "RNA")

raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

library(scran)
library(SingleR)
# 3. Run SingleR
#With only DSS subset of reference
ct_ann <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype,
                  de.method = 'wilcox')
```




```{r}
#Subset to keep only relevant cells
ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(ref.sub)
unique(ref.sub$celltype_integration)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")

unique(ref$celltype_integration)


unique(ref$celltype_general)

raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
```



```{r}
#Stromal.sub$cellType <- ct_ann$labels[match(rownames(Stromal.sub@meta.data), rownames(ct_ann))]
Stromal.sub <- AddMetaData(Stromal.sub, pred$pruned.labels, col.name = 'celltype_integration')
DimPlot(Stromal.sub, reduction = 'umap', group.by = 'celltype_integration')

```

```{r}
library(patchwork)
plot1 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'seurat_clusters')
#plot2 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'cellType')
plot3 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'celltype_integration')
combined_plot <- plot1 | plot3
```



```{r}
plotScoreHeatmap(pred)
# Inspect quality of the predictions
plotDeltaDistribution(pred, ncol = 4, dots.on.top = T)

saveRDS(Stromal.sub, "Stromal_Colon.rds")

```

Cluster 7 has a mixture of the celltypes. Set as unknown?
```{r}
Stromal.sub.markers <- FindMarkers(Stromal.sub, ident.1 = 7, only.pos = TRUE)

Stromal.sub.markers %>%
  #  group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Stromal.sub.markers %>%
   # group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
```
Cluster 7 in "stromal.sub" corresponds to enterocytes (Has the markers and positive for Epcam. Subset out)

```{r}
Enterocytes <- subset(Stromal.sub, idents = 7)
Stromal.sub <- subset(Stromal.sub, idents = 7, invert = TRUE)

# Epithelial
Stromal.sub <- SCTransform(Stromal.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal.sub <- RunPCA(Stromal.sub, verbose = FALSE)
Stromal.sub <- RunUMAP(Stromal.sub, dims = 1:30, verbose = FALSE)

Stromal.sub <- FindNeighbors(Stromal.sub, dims = 1:30, verbose = FALSE)
Stromal.sub <- FindClusters(Stromal.sub, verbose = FALSE, resolution = 0.5)
p1 <- DimPlot(Stromal.sub, label = TRUE)
p2 <- DimPlot(Stromal.sub, label = TRUE, group.by = "celltype_integration")

p1|p2
```

```{r}
raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Stromal.sub$celltype_integration <- pred$pruned.labels[match(rownames(Stromal.sub@meta.data), rownames(pred))]
DimPlot(Stromal.sub, label = TRUE, group.by = "celltype_integration")
```

Pruned labels
```{r}

Stromal.sub <- SetIdent(Stromal.sub, value = "celltype_integration")
DimPlot(Stromal.sub, label = T , repel = T, label.size = 3) + NoLegend()

```

#Using another dataset as reference
```{r}
stroma.ref <-subset(Epi_Imm_Stro_colon, tissueLayer == "stroma")
#ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(stroma.ref)
unique(ref.sub$cellType)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")
raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Stromal.sub$celltype_integration <- pred$pruned.labels[match(rownames(Stromal.sub@meta.data), rownames(pred))]
DimPlot(Stromal.sub, label = TRUE, group.by = "celltype_integration")
```

#Combine both annotations:
```{r}
# Access metadata of the Seurat object
metadata <- Stromal.sub@meta.data
# Create the new column "cellType"
metadata$cellType <- ifelse(metadata$celltype == "ICCajal", metadata$cellType_pruned, metadata$celltype)
# Assign the updated metadata back to the Seurat object
Stromal.sub@meta.data <- metadata

DimPlot(Stromal.sub, label = TRUE, group.by = "cellType")
saveRDS(Stromal.sub, "Stromal_Colon.rds")

```



```{r}
#Sfrp1 and Grem1  markers of CD81+ fibroblasts. Cd81 (Crypt FB)
CD81_Fibroblasts <- c("Cd81", "Sfrp1", "Grem1", "Wnt2b", "Rspo3", "Ackr4")


# PDGFRαhi Fibroblasts (Telocytes, Crypt-Top Fibroblasts, Ednrb hi Fibroblasts)
#PDGFRαhi fibroblasts: "Bmp3", "Bmp7" (Crypt top)

PDGFRalpha_hi_Fibroblasts <- c( "Bmp3", "Bmp7", "Wnt5a", "F3", "Sox6", "Foxl1")

# PDGFRαloCD81- Fibroblasts
PDGFRalpha_lo_CD81_neg_Fibroblasts <- c("Col15a1", "Igfbp5", "Cd90", "Fgfr2", "Fbln")

# Pericytes
Pericytes <- c("Rgs5")

# Smooth Muscle Cells (SMCs)
Smooth_Muscle_Cells <- c("Acta2", "Myh11", "Des")

# Myofibroblasts
Myofibroblasts <- c("Acta2", "Myh11")  # Note: Desmin is also variable

LEC <-c("Lyve1", "Prox1")

Glial <- c("Gfap", "Fabp7", "S100b")

Endothelial <- "Pecam1"

ICC<- c("Kit", "Ano1")
```



To annotate epithelial cell types using SingleR. Use the cell type annotations from :obj.Epithelial.LP_filtered (EPITHELIAL STROMAL AND IMMUNE). As reference.

Add enterocytes to the subset
```{r}
Epithelial.sub <- merge(Epithelial.sub, Enterocytes)
Epithelial.sub <- SCTransform(Epithelial.sub)
# These are now standard steps in the Seurat workflow for visualization and clustering
Epithelial.sub <- RunPCA(Epithelial.sub, verbose = FALSE)
Epithelial.sub <- RunUMAP(Epithelial.sub, dims = 1:30, verbose = FALSE)

Epithelial.sub <- FindNeighbors(Epithelial.sub, dims = 1:30, verbose = FALSE)
Epithelial.sub <- FindClusters(Epithelial.sub, verbose = FALSE, resolution = 0.4)
DimPlot(Epithelial.sub, label = TRUE)

```


Reference but for endothelial cells
```{r}
Epithelial.ref <-subset(obj.Epithelial.LP_filtered, tissueLayer == "epithelium")

#celltype general
# Create a new column 'celltype_general' with generalized cell types
Epithelial.ref@meta.data <- Epithelial.ref@meta.data %>%
  mutate(celltype_general = case_when(
    cellType %in% c("Goblet1", "Goblet2") ~ "Goblet",
    cellType %in% c("AltEnterocyte", "AltEnterocyte2", "AltEnterocyte3", "Enterocyte", "ImmEntero1", "ImmEntero2") ~ "Enterocyte",
    cellType %in% c("TA1", "TA2") ~ "TA",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))


#ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(Epithelial.ref)
unique(ref.sub$celltype_general)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")

```


```{r}
library(scran)
library(SingleR)

counts <- LayerData(Epithelial.sub, assay = "SCT") 

# 3. Run SingleR
ct_ann <- SingleR(test = counts, # we could also use sce or raw_counts
                  ref = ref, 
                  labels = ref$celltype_general,
                  de.method = 'wilcox')
```

```{r}
Epithelial.sub <- AddMetaData(Epithelial.sub, ct_ann$pruned.labels, col.name = 'celltype_integration')
DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'celltype_integration', label = TRUE)

library(patchwork)
plot1 <- DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'seurat_clusters')
#plot2 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'cellType')
plot3 <- DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'celltype_integration')
combined_plot <- plot1 | plot3


saveRDS(Epithelial.sub, "Epithelial_Colon")
```

Evaluation Diagnostics
```{r}
plotScoreHeatmap(ct_ann)

```


```{r}
# Inspect quality of the predictions
plotScoreHeatmap(pred)
plotDeltaDistribution(pred, ncol = 4, dots.on.top = T)
```
Try using normalized counts
```{r}
norm_counts <- LayerData(Stromal_ctrl_DSS, assay = "SCT", layer = 'data') 
ref <- obj.Epithelial.LP_filtered

Idents(ref) <- "cellType"
ref <- subset(ref, idents = c("ICCajal", "Glia1", "Glia2", "TA1", "TA2", "StemCell",
                         "AltEnteroPC", "EnteroPrecur", "EnteroEnd1", "EnteroEnd2",
                         "Bcell1_IgM", "Bcell2_IgM", "Plasma_IgA", "TcellCD8_NK",
                         "Tcell2", "Dendritic", "ActDendri", "Macrophage",
                         "InjuryMono1", "InjuryMono2", "ActNeutrophil", "LymphEndo"), invert = TRUE)

# Create a new column 'celltype_general' with generalized cell types
ref@meta.data <- ref@meta.data %>%
  mutate(celltype_general = case_when(
    cellType %in% c("Goblet", "Goblet1", "Goblet2") ~ "Goblet",
    cellType %in% c("AltEnterocyte", "AltEnterocyte2", "AltEnterocyte3", "Enterocyte", "ImmEntero1", "ImmEntero2") ~ "Enterocyte",
    cellType %in% c("BaseCryptFB1", "InjuryCryptFB1", "InjuryCryptFB2", "BaseCryptFB2", "CryptFB1", "CryptFB2") ~ "CryptFB",
    cellType %in% c("VascSmMuscle1", "VascSmMuscle2") ~ "VascSmMuscle",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))

ref<- as.SingleCellExperiment(ref, assay = "SCT")

pred <- SingleR(test = norm_counts, # we could also use sce or raw_counts
                  ref = ref, 
                  labels = ref$celltype_general,
                  de.method = 'wilcox')
```


```{r}
Stromal_ctrl_DSS$cellType <- ct_ann$labels[match(rownames(Stromal_ctrl_DSS@meta.data), rownames(ct_ann))]

Stromal_ctrl_DSS$pred <- pred$labels[match(rownames(Stromal_ctrl_DSS@meta.data), rownames(pred))]

  
DimPlot(Stromal_ctrl_DSS, reduction = 'umap', group.by = 'cellType')
DimPlot(Stromal_ctrl_DSS, reduction = 'umap', group.by = 'pred', label = "TRUE")

```

Evaluation Diagnostics
```{r}
#ct_ann
#correlation of each cell to all the labels. 
#ct_ann$scores
plotScoreHeatmap(ct_ann)
```


```{r}
# Inspect quality of the predictions
plotScoreHeatmap(pred)
plotDeltaDistribution(pred, ncol = 4, dots.on.top = T)
```


#In reference:
Join goblet, goblet 1 and goblet 2
Join AltEntrocyte, AltEntrocyte2, AltEntrocyte3 and Enterocyte as well as immentero1 and immentero2
Join all crypt cell types
mural cell fibroblasts can stay as they are as well as myofb and FB3

```{r}

```


```{r}
# Add to seurat object
rownames(ct_ann)[1:5] # make sure you have cell IDs
seu <- AddMetaData(seu, ct_ann$pruned.labels, col.name = 'SingleR_HCA')

# Visualise them on the UMAP
seu <- SetIdent(seu, value = "SingleR_HCA")
DimPlot(seu, label = T , repel = T, label.size = 3) + NoLegend()

```


Annotate cell types using GPTCellType

```{r}
# IMPORTANT! Assign your OpenAI API key. See Vignette for details
Sys.setenv(OPENAI_API_KEY = 'your_openai_API_key')

# Load packages
library(GPTCelltype)
library(openai)

# Assume you have already run the Seurat pipeline https://satijalab.org/seurat/
# "obj" is the Seurat object; "markers" is the output from FindAllMarkers(obj)
# Cell type annotation by GPT-4
res <- gptcelltype(stromal.markers, model = 'gpt-4')

# Assign cell type annotation back to Seurat object
obj@meta.data$celltype <- as.factor(res[as.character(Idents(obj))])

# Visualize cell type annotation on UMAP
DimPlot(obj,group.by='celltype')
Epithelial.LP.ctrl$Treatment <- obj$condition
```



Sample 5.
##Fibroblasts
```{r}
df.ctrl <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_DSS_expression_matrix.txt")
df.ctrl <- df.ctrl %>%
    column_to_rownames(var = "V1")

head(df.ctrl)

df.dss <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_HC_expression_matrix.txt")
df.dss <- df.dss %>%
    column_to_rownames(var = "V1")

Fibroblasts.ctrl <- CreateSeuratObject(df.ctrl, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

Fibroblasts.dss <- CreateSeuratObject(df.dss, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

```
QC on sample on its own

```{r}
Fibroblasts.ctrl[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.ctrl, pattern = "mt-")
Fibroblasts.dss[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.dss, pattern = "mt-")

#CTRL
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.ctrl <- PercentageFeatureSet(Fibroblasts.ctrl, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.ctrl <- subset(Fibroblasts.ctrl, subset = nFeature_RNA > 300 & nFeature_RNA < 5000)


#DSS
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

p# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.dss <- PercentageFeatureSet(Fibroblasts.dss, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.dss <- subset(Fibroblasts.dss, subset = nFeature_RNA > 300 & nFeature_RNA < 4000)
```


Downstream analysis. What Cell Types?
```{r}
# run sctransform
Fibroblasts.ctrl <- SCTransform(Fibroblasts.ctrl, vars.to.regress = "percent_ribo", verbose = FALSE, return.only.var.genes = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Fibroblasts.ctrl <- RunPCA(Fibroblasts.ctrl, verbose = FALSE)
Fibroblasts.ctrl <- RunUMAP(Fibroblasts.ctrl, dims = 1:20, verbose = FALSE)

Fibroblasts.ctrl <- FindNeighbors(Fibroblasts.ctrl, dims = 1:20, verbose = FALSE)
Fibroblasts.ctrl <- FindClusters(Fibroblasts.ctrl, verbose = FALSE, resolution = 0.5)
DimPlot(Fibroblasts.ctrl, label = TRUE)

# run sctransform
Fibroblasts.dss <- SCTransform(Fibroblasts.dss, vars.to.regress = "percent_ribo", verbose = FALSE,return.only.var.genes = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Fibroblasts.dss <- RunPCA(Fibroblasts.dss, verbose = FALSE)
Fibroblasts.dss <- RunUMAP(Fibroblasts.dss, dims = 1:20, verbose = FALSE)

Fibroblasts.dss <- FindNeighbors(Fibroblasts.dss, dims = 1:20, verbose = FALSE)
Fibroblasts.dss <- FindClusters(Fibroblasts.dss, verbose = FALSE, resolution = 0.5)
DimPlot(Fibroblasts.dss, label = TRUE)
```

Annotate cells
```{r}
ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(ref.sub)
unique(ref.sub$celltype_general)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")

raw_counts <- LayerData(Fibroblasts.ctrl, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Fibroblasts.ctrl <- AddMetaData(Fibroblasts.ctrl, pred$pruned.labels, col.name = 'celltype_integration_2')
DimPlot(Fibroblasts.ctrl, label = TRUE, group.by = "celltype_integration_2")
saveRDS(Fibroblasts.ctrl, "Fibroblasts_ctrl.rds")

stroma.ref <-subset(Epi_Imm_Stro_colon, tissueLayer == "stroma")
#ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(stroma.ref)
unique(ref.sub$cellType)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")



raw_counts <- LayerData(Fibroblasts.dss, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Fibroblasts.dss$celltype <- pred$labels[match(rownames(Fibroblasts.dss@meta.data), rownames(pred))]
Fibroblasts.dss <- AddMetaData(Fibroblasts.dss, pred$pruned.labels, col.name = 'celltype_integration')
DimPlot(Fibroblasts.dss, label = TRUE, group.by = "celltype_integration")
saveRDS(Fibroblasts.dss, "Fibroblasts_dss.rds")


```







#Manually annotating cells according to markers

"Pecam1",  "Kdr","Plvap <- BEC
"Col1a1", "Col3a1","Fn1","Dcn", "Loxl1", "Loxl2",  "Mmp2"<- Fibroblasts
"Gfap", "Fabp7", "S100b" <- Glial
"Ano1", "Kit" <- ICC
"Lyve1", "Prox1" <- LEC
Rgs5<- Pericytes
"Acta2", "Cnn1", "Mylk"<- SMC


```{r}
BEC <- c("Pecam1","Kdr","Plvap")
Fibroblasts <- c("Col1a1", "Col3a1","Fn1","Dcn", "Loxl1", "Loxl2",  "Mmp2")
Glial <- c("Gfap", "Fabp7", "S100b")
ICC <- c("Ano1", "Kit")
LEC <- c("Lyve1", "Prox1")
SMC <- c("Acta2", "Cnn1", "Mylk")

Fibroblast_markers<- c("Acta2", "Dpt", "Col5a1", "Col1a2")
ICC <- c("Ano1", "Kit")
SM <- c("Myh11", "Des")
Pericytes <- c("Rgs5", "Pdgfrb")
Endothelial <- "Pecam1"
Immune <- c("Cd52", "Ptprc")
Glial <- c("S100b", "Gfap")




```




Clustering cells from healthy mice revealed 13 distinct clusters (Figure 3A and 3C ). We readily identified clusters showing specific expression of epithelial (Epcam and Krt19), pericyte (Rgs5 and Pdgfrb), vascular endothelial (Pecam1 / Cd31), lymphatic endothelial (Lyve1), and glial (S100b and Gfap) and hematopoietic cell markers (Cd52 and Ptprc / Cd45) (Figure 3C). Cluster 2, a small cluster of 32 cells, expressed markers associated with enteric smooth muscle (Myh11 and Des) and interstitial cells of Cajal (ICCs) (Kit and Ano1). Further examination of this cluster revealed its composition was two distinct sub-clusters consistent with ICCs and smooth muscle cells, respectively (Figure 3A). Other low-abundance clusters included enteric glial cells (14 cells) and pericytes (67 cells) (Figure 3A). The remaining 6 cell clusters (4–5, 10–13), comprising 3,391 cells or 89% of the dataset were fibroblast-like cells (FLCs) characterized by expression of the pan-fibroblast markers such as Dpt, Col6a2, and Col1a2 (Figure 3C). Clusters 4 and 5 also showed α-Sma expression, while only cluster 4 showed significant expression of smooth muscle myosin (Myh11) (Figure 3C). We readily identified these six populations as putative counterparts to the stromal cell populations in our human data by cluster marker expression (Table S6).



From:https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9399414/

PDGFRαhi fibroblasts: "Bmp3", "Bmp7" (Crypt top)


Pdgfra-NG2+Rgs5 + pericytes: "Rgs5"

Varying levels of Acta2, Myh11 and Des can help with the distinction between SMCs and myofibroblasts, but the two terms are sometimes used interchangeably in single cell RNA sequencing analyses (9–12, 14, 16, 18). Notably, the small intestine and colon display similar mesenchymal subsets with location-specific differences in their transcriptional profiles (5, 10).

SMC <- c("Acta2", "Cnn1", "Mylk")
SM <- c("Myh11", "Des")

Fibroblast_markers<- c("Acta2", "Dpt", "Col5a1", "Col1a2")

Sfrp1 and Grem1  markers of CD81+ fibroblasts. Cd81 (Crypt FB)
 PDGFRαloCD81- Fibroblasts

```{r}
FeaturePlot(Fibroblasts.ctrl, features = c("Kit", "Ano1"))
FeaturePlot(Fibroblasts.dss, features = c("Kit", "Ano1"))

FeaturePlot(Fibroblasts.dss, features = c("Dpt", "Col6a2", "Col1a2"))
#Dpt, Col6a2, and Col1a2 

FeaturePlot(Fibroblasts.dss, features = BEC)

```


```{r}
ctrl.markers <- FindAllMarkers(Fibroblasts.ctrl, only.pos = TRUE)
ctrl.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

ctrl.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10.ctrl
DoHeatmap(Fibroblasts.ctrl, features = top10.ctrl$gene) + NoLegend()


dss.markers <- FindAllMarkers(Fibroblasts.dss, only.pos = TRUE)
dss.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

dss.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10.dss
DoHeatmap(Fibroblasts.dss, features = top10.dss$gene) + NoLegend()

```





```{r}
Fibroblasts <- merge(Fibroblasts.ctrl, Fibroblasts.dss, add.cell.ids = c("ctrl", "DSS"))

#Fibroblasts <- list(Fibroblasts.ctrl, Fibroblasts.dss)

Fibroblasts$Treatment <- sub("_.*", "", colnames(Fibroblasts))

obj <- Fibroblasts

#obj[["RNA"]] <- split(obj[["RNA"]], f = obj$Treatment)

obj <- NormalizeData(Fibroblasts)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

ElbowPlot(obj)

obj <- FindNeighbors(obj, dims = 1:20, reduction = "pca")
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:20, reduction = "pca", reduction.name = "umap.unintegrated")
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("Treatment"))

obj <- IntegrateLayers(
  object = obj, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:20)
obj <- FindClusters(obj, resolution = 0.5, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:20, reduction.name = "umap.harmony")
p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("harmony_clusters", "Treatment"),
  combine = FALSE, label.size = 2, label = TRUE
)
library(patchwork)
 wrap_plots(c(p1), ncol = 2, nrow = 2)
 

```
#Paper, why are the treatment ans ctrl so different? 
Strong batch effect

Another approach
```{r}

df.ctrl <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_DSS_expression_matrix.txt")
df.ctrl <- df.ctrl %>%
    column_to_rownames(var = "V1")

head(df.ctrl)

df.dss <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_HC_expression_matrix.txt")
df.dss <- df.dss %>%
    column_to_rownames(var = "V1")

Fibroblasts.ctrl <- CreateSeuratObject(df.ctrl, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

Fibroblasts.dss <- CreateSeuratObject(df.dss, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

Fibroblasts.ctrl[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.ctrl, pattern = "mt-")
Fibroblasts.dss[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.dss, pattern = "mt-")

#CTRL
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.ctrl <- PercentageFeatureSet(Fibroblasts.ctrl, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.ctrl <- subset(Fibroblasts.ctrl, subset = nFeature_RNA > 300 & nFeature_RNA < 5000)


#DSS
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

p# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.dss <- PercentageFeatureSet(Fibroblasts.dss, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.dss <- subset(Fibroblasts.dss, subset = nFeature_RNA > 300 & nFeature_RNA < 4000)

```

```{r}



Fibroblasts <- merge(Fibroblasts.ctrl, Fibroblasts.dss, add.cell.ids = c("ctrl", "DSS"))

#Fibroblasts <- list(Fibroblasts.ctrl, Fibroblasts.dss)

Fibroblasts$Treatment <- sub("_.*", "", colnames(Fibroblasts))

Fibroblasts <- JoinLayers(Fibroblasts)

alldata.list <- SplitObject(Fibroblasts, split.by = "Treatment")

for (i in 1:length(alldata.list)) {
    alldata.list[[i]] <- NormalizeData(alldata.list[[i]], verbose = FALSE)
    alldata.list[[i]] <- FindVariableFeatures(alldata.list[[i]], selection.method = "vst", nfeatures = 2000,verbose = FALSE)
}

# get the variable genes from all the datasets.
hvgs_per_dataset <- lapply(alldata.list, function(x) { x@assays$RNA@var.features })


# also add in the variable genes that was selected on the whole dataset
hvgs_per_dataset$all = VariableFeatures(Fibroblasts)

temp <- unique(unlist(hvgs_per_dataset))
overlap <- sapply( hvgs_per_dataset , function(x) { temp %in% x } )
pheatmap::pheatmap(t(overlap*1),cluster_rows = F ,
                   color = c("grey90","grey20"))
```



#Gene markers
```{r}
markers <- c("Pecam1",  "Kdr","Plvap","Col1a1", "Col3a1","Fn1","Dcn", "Loxl1", "Loxl2",  "Mmp2", "Gfap", "Fabp7", "S100b", "Gob",
       "Kit",  "Ano1", "Lyve1", "Prox1", "Pdgfra", "Smoc2", "Rgs5", "Acta2", "Cnn1", "Mylk")
```
"Pecam1",  "Kdr","Plvap <- BEC
"Col1a1", "Col3a1","Fn1","Dcn", "Loxl1", "Loxl2",  "Mmp2"<- Fibroblasts
"Gfap", "Fabp7", "S100b" <- Glial
"Ano1", "Kit" <- ICC
"Lyve1", "Prox1" <- LEC
Rgs5<- Pericytes
"Acta2", "Cnn1", "Mylk"<- SMC


Clustering cells from healthy mice revealed 13 distinct clusters (Figure 3A and 3C ). We readily identified clusters showing specific expression of epithelial (Epcam and Krt19), pericyte (Rgs5 and Pdgfrb), vascular endothelial (Pecam1 / Cd31), lymphatic endothelial (Lyve1), and glial (S100b and Gfap) and hematopoietic cell markers (Cd52 and Ptprc / Cd45) (Figure 3C). Cluster 2, a small cluster of 32 cells, expressed markers associated with enteric smooth muscle (Myh11 and Des) and interstitial cells of Cajal (ICCs) (Kit and Ano1
```{r}
# CD81+ Fibroblasts (Trophocytes, Crypt-Bottom Fibroblasts, MRISCs, pi16+ Fibroblasts)
#
#Sfrp1 and Grem1  markers of CD81+ fibroblasts. Cd81 (Crypt FB)
CD81_Fibroblasts <- c("Cd81", "Sfrp1", "Grem1", "Wnt2b", "Rspo3", "Ackr4")


# PDGFRαhi Fibroblasts (Telocytes, Crypt-Top Fibroblasts, Ednrb hi Fibroblasts)
#PDGFRαhi fibroblasts: "Bmp3", "Bmp7" (Crypt top)

PDGFRalpha_hi_Fibroblasts <- c( "Bmp3", "Bmp7", "Wnt5a", "F3", "Sox6", "Foxl1")

# PDGFRαloCD81- Fibroblasts
PDGFRalpha_lo_CD81_neg_Fibroblasts <- c("Col15a1", "Igfbp5", "Cd90", "Fgfr2", "Fbln")

# Pericytes
Pericytes <- c("Rgs5")

# Smooth Muscle Cells (SMCs)
Smooth_Muscle_Cells <- c("Acta2", "Myh11", "Des")

# Myofibroblasts
Myofibroblasts <- c("Acta2", "Myh11")  # Note: Desmin is also variable

LEC <-c("Lyve1", "Prox1")

Glial <- c("Gfap", "Fabp7", "S100b")

Endothelial <- "Pecam1"

ICC<- c("Kit", "Ano1")

```


```{r}
Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= CD81_Fibroblasts,
                                 name = "CD81_Fibroblasts")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= PDGFRalpha_hi_Fibroblasts,
                                 name = "PDGFRalpha_hi_Fibroblasts")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= c("Col15a1", "Igfbp5", "Fgfr2"),
                                 name = "PDGFRalpha_lo_CD81_neg_Fibroblasts")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= Pericytes,
                                 name = "Pericytes")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= Smooth_Muscle_Cells,
                                 name = "Smooth_Muscle_Cells")


Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= LEC,
                                 name = "LEC")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= Glial,
                                 name = "Glial")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= Endothelial,
                                 name = "Endothelial")

Fibroblasts.ctrl <- AddModuleScore(object=Fibroblasts.ctrl,
                                 features= ICC,
                                 name = "ICC")
```


Plot

```{r}
seu.int <- AddModuleScore(seu.int, features = sig.list, name = "_score")
names(seu.int@meta.data)[grep("_score", names(seu.int@meta.data))] <- names(sig.list)
FeaturePlot(seu.int, features = names(sig.list))
```


```{r}
library(RColorBrewer)

# Plot scores
FeaturePlot(Fibroblasts.ctrl,
            features = Endothelial, label = TRUE, repel = TRUE) +
            scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))

# Plot scores
FeaturePlot(Fibroblasts.ctrl,
            features = c("CD81_Fibroblasts1","PDGFRalpha_hi_Fibroblasts1", "PDGFRalpha_lo_CD81_neg_Fibroblasts1"), label = TRUE, repel = TRUE) +
            scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))


```



```{r}
g <- c("Pdpn", "Cd81","Pcolce2","C3", "Cd55","Grem1","Agt", "Lpl", "Procr",  "Acta2", "Adamdec1", "Plvap", "Ano1", 
       "Kit",  "Ptn", "Cnn1", "Mylk", "Rgs5", "Lyve1", "Cd52", "Plp1", "Epcam")

markers <- c("Pdpn",  "Cd81","Pcolce2","C3", "Cd55","Grem1","Agt", "Lpl", "Procr",  "Acta2", "Adamdec1", "Plvap", "Ano1", 
       "Kit",  "Ptn", "Cnn1", "Mylk", "Rgs5", "Lyve1", "Cd52", "Plp1", "Epcam")

DotPlot(all.anchors.cluster, g, cols =c("midnightblue", "darkorange1")) + theme(axis.text.x = element_text(angle = 70, hjust = 1))


```

```{r}
FeaturePlot(Fibroblasts.ctrl, features = markers)
```


```{r}

```


```{r}


#create a list of Seurat objects and perform preprocessing
data.list <- lapply(X = list(Fibroblasts.ctrl, Fibroblasts.dss), FUN = function(x) {
# making a mitochondria percentage column (mt/MT/Mt based on gtf)
# Filtering, normalization and feature detection
x <- NormalizeData(x)
x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})


# find integration anchors ...
pancreas.anchors <- FindIntegrationAnchors(object.list = data.list, dims = 1:30)
# ... and integrate
pancreas.integrated <- IntegrateData(anchorset = pancreas.anchors)
# switch to integration assay
DefaultAssay(pancreas.integrated) <- "integrated"

# Standard workflow for visualization and clustering
pancreas.integrated <- ScaleData(pancreas.integrated, verbose = FALSE) 
pancreas.integrated <- RunPCA(pancreas.integrated, assay = "integrated", npcs = 30, verbose = FALSE) 
pancreas.integrated <- RunUMAP(pancreas.integrated, reduction = "pca", dims = 1:30, verbose = FALSE)

pancreas.integrated <- FindNeighbors(data.combined, reduction = "pca", dims = 1:15) 
pancreas.integrated <- FindClusters(data.combined, resolution = 0.5)

# Number of cells in each condition
table(pancreas.integrated@meta.data$tech) table(pancreas.integrated@meta.data$tech, pancreas.integrated@meta.data$celltype)
# Visualization
p1 <- DimPlot(pancreas.integrated, reduction = "umap", group.by = "Treatment")
p2 <- DimPlot(pancreas.integrated, reduction = "umap", group.by = "celltype", label = TRUE, repel = TRUE) + NoLegend()
p3 <- DimPlot(pancreas.integrated, reduction = "umap", split.by = " tech")
p1 + p2 + p3
```

