---
title: "Individual_datasets_integration"
output: html_document
date: "2024-07-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Sample: Colonic Stroma: GSE172261

DSS+WT

Time: 7 day cycles * 3 = 21 days

Mice were subjected to 3 repetitive cycles of DSS displayed progressive accumulation of immune cell infiltrates associated with excessive deposition of collagen fibers. Lamina propria cells from water-fed and DSS-fed mice were isolated using enzymatic digestion, and enriched for stromal cells by FACS using antibodies excluding hematopoietic cells (CD45), epithelial cells (EpCAM), and erythrocytes. Prepared single cell suspensions were then profiled using the 10x Chromium V2 droplet-based single cell RNA sequencing platform. Tissues were collected from proximal and distal colon.


```{r}
library(Seurat)
library(readr)

Colonic_Stroma_DSS <- ReadMtx(
  mtx = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_matrix.mtx", features = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_genes.tsv",
  cells = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_barcodes.tsv"
)

Colonic_Stroma_DSS <- CreateSeuratObject(counts = Colonic_Stroma_DSS, min.cells = 3, min.features = 200)

# Reading metadat
metadata<-readr::read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_Stroma/GSE172261_metadata.tsv")
print(metadata)

library(dplyr)
library(tibble)
# Set row names of metadata to cell_barcode
metadata <- metadata %>%
  tibble::column_to_rownames(var = "cell_barcode")

# Ensure the barcodes in metadata match the Seurat object
common_barcodes <- intersect(rownames(metadata), colnames(Colonic_Stroma_DSS))

# Subset metadata to only include common barcodes
metadata <- metadata[common_barcodes, ]

# Add metadata to Seurat object
Colonic_Stroma_DSS <- AddMetaData(Colonic_Stroma_DSS, metadata)

#Quality Control
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Colonic_Stroma_DSS[["percent.mt"]] <- PercentageFeatureSet(Colonic_Stroma_DSS, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Colonic_Stroma_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Colonic_Stroma_DSS, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Colonic_Stroma_DSS, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Colonic_Stroma_DSS@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Colonic_Stroma_DSS <- PercentageFeatureSet(Colonic_Stroma_DSS, "^Rp", col.name = "percent_ribo")
VlnPlot(Colonic_Stroma_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Colonic_Stroma_DSS <- subset(Colonic_Stroma_DSS, subset = nFeature_RNA > 300 & nFeature_RNA < 3400 & percent.mt < 8 & percent_ribo <25)

View(Colonic_Stroma_DSS@meta.data)

#Create one column with all the cell types
library(dplyr)

Colonic_Stroma_DSS$celltype <- Colonic_Stroma_DSS@meta.data %>% 
  mutate(celltype = coalesce(Celltype_Stroma_Atlas,Celltype_Endothelium_Atlas,Celltype_Fibroblast_Atlas))

Colonic_Stroma_DSS$Age <- "12 weeks"

obj <- Colonic_Stroma_DSS

obj[["RNA"]] <- split(obj[["RNA"]], f = obj$batch)
obj

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("batch", "celltype"))

obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj, resolution = 0.5, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")
p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("batch", "celltype", "harmony_clusters", "Treatment"),
  combine = FALSE, label.size = 2, label = TRUE
)
library(patchwork)
 wrap_plots(c(p1), ncol = 2, nrow = 2)
 
 DimPlot(obj, reduction = "umap.harmony", group.by = "Site", combine = FALSE, label.size = 2, label = TRUE
)

obj_Colonic_Stroma_DSS <- obj

# Replace "H2O" with "ctrl" in the Treatment metadata of the Seurat object
obj_Colonic_Stroma_DSS$Treatment <- ifelse(obj_Colonic_Stroma_DSS$Treatment == "H2O", "ctrl", obj_Colonic_Stroma_DSS$Treatment)

obj_Colonic_Stroma_DSS$cellType <- obj_Colonic_Stroma_DSS$celltype

# Create a new column 'celltype_general' with generalized cell types
obj_Colonic_Stroma_DSS@meta.data <- obj_Colonic_Stroma_DSS@meta.data %>%
  mutate(celltype_general = case_when(
    cellType %in% c("Fibroblast 1a", "Fibroblast 1b", "Fibroblast 1c") ~ "Fibroblast 1",
    cellType %in% c("SMC 1", "SMC 2") ~ "SMC",
    cellType %in% c("Fibroblast 2c", "Fibroblast 2a", "Fibroblast 2b") ~ "Fibroblast 2",
    cellType %in% c("Fibroblast 3a", "Fibroblast 3b") ~ "Fibroblast 3",
    cellType %in% c("BEC 1", "BEC 2") ~ "BEC",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))

# Create a new column 'celltype_general' with generalized cell types
library(dplyr)
obj_Colonic_Stroma_DSS@meta.data <- obj_Colonic_Stroma_DSS@meta.data %>%
  mutate(celltype_integration = case_when(
    cellType %in% c("Fibroblast 1a", "Fibroblast 1b", "Fibroblast 1c", "Fibroblast 2c", "Fibroblast 2a", "Fibroblast 2b", "Fibroblast 1","Fibroblast 3a", "Fibroblast 3b") ~ "Fibroblast",
    cellType %in% c("SMC 1", "SMC 2") ~ "SMC",
    cellType %in% c("BEC 1", "BEC 2") ~ "BEC",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))

#Fibroblast 1 is CD81 pos
#Fibroblast 2 is CD81 pos ->PDGFRalpha_lo_CD81_neg_Fibroblasts
#Fibroblast 3 is CD81 pos ->PDGFRalpha_hi_Fibroblasts

#How was the annotation of the dataset given?
#https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3001532 

#Check cell cycle scores
#Convert Human genes to Mouse
library(gprofiler2)
mmus_s = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

obj_Colonic_Stroma_DSS_joint <- JoinLayers(obj_Colonic_Stroma_DSS)

obj_Colonic_Stroma_DSS <- CellCycleScoring(
    object = obj_Colonic_Stroma_DSS_joint,
    g2m.features = mmus_g2m,
    s.features = mmus_s
)

VlnPlot(obj_Colonic_Stroma_DSS, features = c("S.Score", "G2M.Score"), group.by = "orig.ident", ncol = 3, pt.size = .1)
FeatureScatter(obj_Colonic_Stroma_DSS, "S.Score", "G2M.Score", group.by = "Phase")

#No need to regress out cell cycle scores

#Check for doublets
library("scDblFinder")
library(SingleCellExperiment)

#scDblFinder works with SCE objects
sce <- as.SingleCellExperiment(obj_Colonic_Stroma_DSS)

#Store the highly variable features as top.var to use with scDblFinder
top.var <- VariableFeatures(obj_Colonic_Stroma_DSS)

dbl.dens <- computeDoubletDensity(sce, subset.row=top.var, 
                                  d=ncol(reducedDim(sce)))
summary(dbl.dens)

sce$DoubletScore <- dbl.dens

seu <- as.Seurat(sce)
#Plot Doublets
FeaturePlot(seu, features = "DoubletScore", reduction = "UMAP.HARMONY")

dbl.calls <- doubletThresholding(data.frame(score=dbl.dens),
                                 method="griffiths", returnType="call")
summary(dbl.calls)
```
Before quality control: 16878 34197
After quality control:

```{r}
saveRDS(obj_Colonic_Stroma_DSS, file = "Colonic_Stroma.rds")
```

What do each of the fibroblast groups correspond to?
```{r}
# Remember to switch to raw data for DEG
DefaultAssay(Colonic_Stroma) <- "RNA"
# Find DEGs in celseq
#celseq.markers <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 1",group.by = "tech", logfc.threshold = 0.25, only.pos = TRUE)

# Find DEGs in alpha
Idents(Colonic_Stroma) <- "celltype_general"
markers.fb1 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 1", logfc.threshold = 0.25, only.pos = TRUE)
markers.fb2 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 2", logfc.threshold = 0.25, only.pos = TRUE)
markers.fb3 <- FindMarkers(Colonic_Stroma, ident.1 = "Fibroblast 3", logfc.threshold = 0.25, only.pos = TRUE)


# Find conserved DEGs among techs
all.markers <- FindConservedMarkers(markers.fb1, ident.1 = c("Fibroblast 1"), grouping.var = "Treatment")
```

Sample 6. Epithelial LP
We first compared the epithelial healing effects of RSPO2 and a Wnt mimetic with broad Fzd-specificity in an acute Dextran Sodium Sulfate (DSS) mouse colitis model. Guided by Fzd expression patterns in the colon epithelium, we also examined the effects of Wnt mimetics with sub-family Fzd-specificities. 
 
Read in the matrices for the wild type samples

```{r}
count_table <- read.table("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Epithelial_LP/GSE201723_SRZ_colon_mat.tsv")
Epithelial.LP <- CreateSeuratObject(counts = count_table, min.cells = 3, min.features = 300)

metadata <- read.table("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Epithelial_LP/GSE201723_SRZ_colon_metaData.tsv")
head(metadata)
Epithelial.LP <- AddMetaData(Epithelial.LP, metadata)

#QC
Epithelial.LP[["percent.mt"]] <- PercentageFeatureSet(Epithelial.LP, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Epithelial.LP, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used

plot1 <- FeatureScatter(Epithelial.LP, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Epithelial.LP, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Epithelial.LP@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

Epithelial.LP <- PercentageFeatureSet(Epithelial.LP, "^Rp", col.name = "percent_ribo")
VlnPlot(Epithelial.LP, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Epithelial.LP <- subset(Epithelial.LP, subset = nFeature_RNA > 300 & nFeature_RNA < 5000 & percent.mt < 4 & percent_ribo <10)
View(Epithelial.LP@meta.data)

```

Keep only uninjured and DSS without treatment. "We first assessed the effect of DSS injury by comparing the DSS, anti-GFP condition with the uninjured condition" 
```{r}
# Subset the Seurat object to keep only cells corresponding to 'Uninj' or 'antiGFP'
Epithelial.LP_filtered <- subset(
  Epithelial.LP, 
  subset = batch %in% c("d5_24h_Uninj_1", "d5_24h_Uninj_2", "d6_48h_Uninj_1", "d6_48h_Uninj_2", "d5_24h_aGFP_1", "d5_24h_aGFP_2", "d5_24h_aGFP_3", "d6_48h_aGFP_1", "d6_48h_aGFP_2", "d6_48h_aGFP_3")
)

Epithelial.LP_filtered$condition <- NA

# Assign values based on the 'batch' column
Epithelial.LP_filtered$condition[grep("Uninj", Epithelial.LP_filtered$batch)] <- "Uninjured"
Epithelial.LP_filtered$condition[grep("aGFP", Epithelial.LP_filtered$batch)] <- "antiGFP"

# Optionally, you can check the unique values of the new 'condition' column
unique(Epithelial.LP_filtered$condition)

obj <- Epithelial.LP_filtered

obj[["RNA"]] <- split(obj[["RNA"]], f = obj$condition)
obj

obj <- SCTransform(obj, vars.to.regress = c("percent.mt", "percent_ribo"), verbose = FALSE)
obj <- RunPCA(obj)

obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 1, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("batch", "cellType"))

Layers(obj)

obj <- Seurat::IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj, resolution = 0.5, cluster.name = "harmony_clusters")
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")

# Replace "antiGFP" with "DSS" in the condition metadata
obj$condition <- ifelse(obj$condition == "antiGFP", "DSS", obj$condition)

p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("batch", "cellType", "harmony_clusters", "condition"),
  combine = FALSE, label.size = 2, label = TRUE
)
library(patchwork)
wrap_plots(c(p1), ncol = 2, nrow = 2)
 
DimPlot(obj, reduction = "umap.harmony", group.by = "tissueLayer", combine = FALSE, label.size = 2, label = TRUE
)
 
DimPlot(obj, reduction = "umap.harmony", group.by = "cellType", combine = FALSE, label.size = 2, label = TRUE
)

obj.Epithelial.LP_filtered <- obj

obj.Epithelial.LP_filtered$Treatment <- obj.Epithelial.LP_filtered$condition

# Replace "H2O" with "ctrl" in the Treatment metadata of the Seurat object
obj.Epithelial.LP_filtered$Treatment <- ifelse(obj.Epithelial.LP_filtered$Treatment == "Uninjured", "ctrl", obj.Epithelial.LP_filtered$Treatment)

 # Create a new column 'celltype_integration' with generalized cell types
library(dplyr)
obj.Epithelial.LP_filtered@meta.data <- obj.Epithelial.LP_filtered@meta.data%>%
mutate(celltype_integration = case_when(
cellType %in% c("Bcell1_IgM", "Bcell2_IgM") ~ "Bcell_IgM",
cellType %in% c("Dendritic", "ActDendri") ~ "DC",
cellType %in% c("InjuryMono1", "InjuryMono2") ~ "InjuryMono",
cellType %in% c("CryptFB1", "CryptFB2", "BaseCryptFB1", "InjuryCryptFB1", "InjuryCryptFB2", "BaseCryptFB2",   "FB3" ) ~ "Fibroblast",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("AltEnteroPC", "EnteroPrecur", "AltEnterocyte", "ImmEntero1", "ImmEntero2", "Enterocyte", "AltEnterocyte2",  "AltEnterocyte3") ~ "Enterocyte",
cellType %in% c("BasalGoblet", "Goblet1", "Goblet2") ~ "Goblet",
cellType %in% c("TA1", "TA2" ) ~ "TA",
cellType %in% c("Glia1", "Glia2" ) ~ "Glia",
TRUE ~ cellType  # Keeps the original cellType for those not specified
))

saveRDS(obj.Epithelial.LP_filtered, file = "Epi_Imm_Stro_colon.rds")

```


Sample 3. IEC_Stroma_RAW GSE163638

Mouse colitis was induced by administration of 2% w/v of dextran sodium sulfate (DSS) dissolved in drinking water ad libitum for 7 days, followed by 7 days of regular water. General health and body weight were monitored regularly. On day 14, colons from untreated and DSS-treated mice were harvested, measured, and processed for spatial transcriptomic studies.

DSS only

To analyze the impact B cells might have on intestinal epithelial cells and stromal cells during recovery after intestinal injury the transcriptional profile of these mice was analysed in mice depleted of B cells and control mice on day 14 after DSS colitis.

Just using ctrl sample. Not cell b depleted.
```{r}
Stromal_ctrl_DSS <- ReadMtx(
  mtx = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-matrix.mtx", features = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-features.tsv",
  cells = "/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/IEC_Stroma_RAW/GSM4983265_IEC-Stroma-control-barcodes.tsv"
)

Stromal_ctrl_DSS <- CreateSeuratObject(counts = Stromal_ctrl_DSS, min.cells = 3, min.features = 200)

#Quality Control
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Stromal_ctrl_DSS[["percent.mt"]] <- PercentageFeatureSet(Stromal_ctrl_DSS, pattern = "mt-")

# Visualize QC metrics as a violin plot
VlnPlot(Stromal_ctrl_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Stromal_ctrl_DSS, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Stromal_ctrl_DSS, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Decide outlier thresholds for on % mito -  manual inspection first 
threshold.mito <- c(2.5,12,15,2.5,10,5,2.5,12.5,12.5,2.5,10,10)

library(ggplot2)
p <- ggplot(Stromal_ctrl_DSS@meta.data, aes(x=percent.mt)) + geom_density(alpha=.2, fill="gray")
p <- p + theme_bw()+ theme(axis.text =element_text(family="Helvetica", face="plain", size=12)) 
p <- p + theme(legend.position="none") + ylab("Density") + xlab("% mitochondrial genes")
p <- p + geom_vline(xintercept=threshold.mito, linetype="dashed", color = "orange", size=1)

p

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Stromal_ctrl_DSS <- PercentageFeatureSet(Stromal_ctrl_DSS, "^Rp", col.name = "percent_ribo")
VlnPlot(Stromal_ctrl_DSS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Stromal_ctrl_DSS <- subset(Stromal_ctrl_DSS, subset = nFeature_RNA > 300 & nFeature_RNA < 4000 & percent.mt < 60 & percent_ribo <55)

View(Stromal_ctrl_DSS@meta.data)


# run sctransform
Stromal_ctrl_DSS <- SCTransform(Stromal_ctrl_DSS, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal_ctrl_DSS <- RunPCA(Stromal_ctrl_DSS, verbose = FALSE)
Stromal_ctrl_DSS <- RunUMAP(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)

Stromal_ctrl_DSS <- FindNeighbors(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)
Stromal_ctrl_DSS <- FindClusters(Stromal_ctrl_DSS, verbose = FALSE, resolution = 0.5)
DimPlot(Stromal_ctrl_DSS, label = TRUE)
Stromal_ctrl_DSS$Treatment <- "DSS"
```


```{r}
# Feature plot - visualize feature expression in low-dimensional space
Idents(Stromal_ctrl_DSS) <- "SCT_snn_res.0.5"

FeaturePlot(Stromal_ctrl_DSS, features = c("Epcam", "Pdgfra", "Ptprc"))

FeaturePlot(Stromal_ctrl_DSS, features = "Pdgfra")
FeaturePlot(Stromal_ctrl_DSS, features = "Ptprc")

```



There shouldn't be immune cells as only stromal and epithelial were gated. Remove cluster 13/clean cells that express ptprc
```{r}
FeaturePlot(Stromal_ctrl_DSS, features = "Ptprc")
#cluster 13
```

```{r}
#subset
Stromal_ctrl_DSS <- subset(Stromal_ctrl_DSS, idents = 13, invert = TRUE)

# run sctransform
Stromal_ctrl_DSS <- SCTransform(Stromal_ctrl_DSS, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal_ctrl_DSS <- RunPCA(Stromal_ctrl_DSS, verbose = FALSE)
Stromal_ctrl_DSS <- RunUMAP(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)

Stromal_ctrl_DSS <- FindNeighbors(Stromal_ctrl_DSS, dims = 1:30, verbose = FALSE)
Stromal_ctrl_DSS <- FindClusters(Stromal_ctrl_DSS, verbose = FALSE, resolution = 0.5)
DimPlot(Stromal_ctrl_DSS, label = TRUE)
```

```{r}
#Subset in epithelial and stromalfor easier annotation
FeaturePlot(Stromal_ctrl_DSS, features = c("Epcam", "Pdgfra", "Ptprc"))
Epithelial.sub <- subset(Stromal_ctrl_DSS, idents = c(0, 2, 12, 7, 8, 9, 10))
Stromal.sub <- subset(Stromal_ctrl_DSS, idents = c(0, 2, 12, 7, 8, 9, 10), invert = TRUE)
```



Annotation
1.Automatic Annotation

```{r}
# Stromal
Stromal.sub <- SCTransform(Stromal.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal.sub <- RunPCA(Stromal.sub, verbose = FALSE)
Stromal.sub <- RunUMAP(Stromal.sub, dims = 1:30, verbose = FALSE)

Stromal.sub <- FindNeighbors(Stromal.sub, dims = 1:30, verbose = FALSE)
Stromal.sub <- FindClusters(Stromal.sub, verbose = FALSE, resolution = 0.4)
DimPlot(Stromal.sub, label = TRUE)
```

```{r}
Stromal.sub.marker <- FindAllMarkers(Stromal.sub, only.pos = TRUE)
Stromal.sub.marker %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Stromal.sub.marker %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(Stromal.sub, features = top10$gene) + NoLegend()
```


```{r}
# Epithelial
Epithelial.sub <- SCTransform(Epithelial.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Epithelial.sub <- RunPCA(Epithelial.sub, verbose = FALSE)
Epithelial.sub <- RunUMAP(Epithelial.sub, dims = 1:30, verbose = FALSE)

Epithelial.sub <- FindNeighbors(Epithelial.sub, dims = 1:30, verbose = FALSE)
Epithelial.sub <- FindClusters(Epithelial.sub, verbose = FALSE, resolution = 0.5)
DimPlot(Epithelial.sub, label = TRUE)

```

```{r}
Epithelial.sub.markers <- FindAllMarkers(Epithelial.sub, only.pos = TRUE)
Epithelial.sub.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Epithelial.sub.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10

DoHeatmap(Epithelial.sub, features = top10$gene) + NoLegend()
```



#Annotate Stromal cells
reference: Colonic Stroma: GSE172261

As the dataset to be annotated contains only DSS stimulated samples, subset the reference
```{r}
library(scran)

ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(ref.sub)
unique(ref.sub$celltype_integration)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")
```


```{r}
library(SingleR)
raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')

Stromal.sub <- AddMetaData(Stromal.sub, pred$pruned.labels, col.name = 'celltype_integration')
DimPlot(Stromal.sub, reduction = 'umap', group.by = 'celltype_integration')
```


```{r}
library(patchwork)
plot1 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'seurat_clusters')
#plot2 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'cellType')
plot3 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'celltype_integration')
combined_plot <- plot1 | plot3
combined_plot
```

Cluster 7 has a mixture of the celltypes. Set as unknown?
```{r}
Stromal.sub.markers <- FindMarkers(Stromal.sub, ident.1 = 7, only.pos = TRUE)

Stromal.sub.markers %>%
  #  group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)

Stromal.sub.markers %>%
   # group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
```
Cluster 7 in "stromal.sub" corresponds to enterocytes (Has the markers and positive for Epcam. Subset out)

```{r}
Enterocytes <- subset(Stromal.sub, idents = 7)
Stromal.sub <- subset(Stromal.sub, idents = 7, invert = TRUE)

# Epithelial
Stromal.sub <- SCTransform(Stromal.sub, vars.to.regress = "percent.mt", verbose = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Stromal.sub <- RunPCA(Stromal.sub, verbose = FALSE)
Stromal.sub <- RunUMAP(Stromal.sub, dims = 1:30, verbose = FALSE)

Stromal.sub <- FindNeighbors(Stromal.sub, dims = 1:30, verbose = FALSE)
Stromal.sub <- FindClusters(Stromal.sub, verbose = FALSE, resolution = 0.5)
p1 <- DimPlot(Stromal.sub, label = TRUE)
p2 <- DimPlot(Stromal.sub, label = TRUE, group.by = "celltype_integration")

p1|p2
```

```{r}
raw_counts <- LayerData(Stromal.sub, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Stromal.sub$celltype_integration <- pred$pruned.labels[match(rownames(Stromal.sub@meta.data), rownames(pred))]
DimPlot(Stromal.sub, label = TRUE, group.by = "celltype_integration")
saveRDS(Stromal.sub, "Stromal_Colon.rds")
```

To annotate epithelial cell types using SingleR. Use the cell type annotations from :obj.Epithelial.LP_filtered (EPITHELIAL STROMAL AND IMMUNE). As reference.

Add enterocytes to the subset
```{r}
Epithelial.sub <- merge(Epithelial.sub, Enterocytes)
Epithelial.sub <- SCTransform(Epithelial.sub)
# These are now standard steps in the Seurat workflow for visualization and clustering
Epithelial.sub <- RunPCA(Epithelial.sub, verbose = FALSE)
Epithelial.sub <- RunUMAP(Epithelial.sub, dims = 1:30, verbose = FALSE)

Epithelial.sub <- FindNeighbors(Epithelial.sub, dims = 1:30, verbose = FALSE)
Epithelial.sub <- FindClusters(Epithelial.sub, verbose = FALSE, resolution = 0.4)
DimPlot(Epithelial.sub, label = TRUE)

```


Reference but for endothelial cells
```{r}
Epithelial.ref <-subset(obj.Epithelial.LP_filtered, tissueLayer == "epithelium")

#celltype general
# Create a new column 'celltype_general' with generalized cell types
Epithelial.ref@meta.data <- Epithelial.ref@meta.data %>%
  mutate(celltype_general = case_when(
    cellType %in% c("Goblet1", "Goblet2") ~ "Goblet",
    cellType %in% c("AltEnterocyte", "AltEnterocyte2", "AltEnterocyte3", "Enterocyte", "ImmEntero1", "ImmEntero2") ~ "Enterocyte",
    cellType %in% c("TA1", "TA2") ~ "TA",
    TRUE ~ cellType  # Keeps the original cellType for those not specified
  ))


#ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(Epithelial.ref)
unique(ref.sub$celltype_general)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")

```


```{r}
library(scran)
library(SingleR)

counts <- LayerData(Epithelial.sub, assay = "SCT") 

# 3. Run SingleR
ct_ann <- SingleR(test = counts, # we could also use sce or raw_counts
                  ref = ref, 
                  labels = ref$celltype_general,
                  de.method = 'wilcox')
```

```{r}
Epithelial.sub <- AddMetaData(Epithelial.sub, ct_ann$pruned.labels, col.name = 'celltype_integration')
DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'celltype_integration', label = TRUE)

library(patchwork)
plot1 <- DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'seurat_clusters')
#plot2 <- DimPlot(Stromal.sub, reduction = 'umap', group.by = 'cellType')
plot3 <- DimPlot(Epithelial.sub, reduction = 'umap', group.by = 'celltype_integration')
combined_plot <- plot1 | plot3
```

```{r}
#Check with known gene markers
# Stem Cells
Stem <- c("Ascl2")
# TA (Transit-Amplifying) Cells
TA <- c("Mki67")
# Goblet Cells
Goblet <- c("Muc2")
# Colonocyte Subpopulations
Colonocyte1 <- c("Car1")
Colonocyte2 <- c("Aqp4")
# EE (Enteroendocrine) Cells
EE <- c("Chga", "Chgb", "Pyy")
# Tuft Cells
Tuft <- c("Fyb", "Dclk1", "Hck")
```

```{r}
FeaturePlot(Epithelial.sub, features = "Ascl2" )
FeaturePlot(Epithelial.sub, features = TA )
FeaturePlot(Epithelial.sub, features = Goblet )
FeaturePlot(Epithelial.sub, features = Colonocyte1 )
FeaturePlot(Epithelial.sub, features = Colonocyte2 )
FeaturePlot(Epithelial.sub, features = EE )
FeaturePlot(Epithelial.sub, features = Tuft )
```
Annotations match to gene cell markers

```{r}
saveRDS(Epithelial.sub, "Epithelial_Colon")
```


Evaluation Diagnostics
```{r}
plotScoreHeatmap(ct_ann)

```

Sample 5.
##Fibroblasts
```{r}
df.ctrl <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_DSS_expression_matrix.txt")
df.ctrl <- df.ctrl %>%
    column_to_rownames(var = "V1")

head(df.ctrl)

df.dss <- data.table::fread("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Fibroblasts/GSE114374_Mouse_HC_expression_matrix.txt")
df.dss <- df.dss %>%
    column_to_rownames(var = "V1")

Fibroblasts.ctrl <- CreateSeuratObject(df.ctrl, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

Fibroblasts.dss <- CreateSeuratObject(df.dss, project = "Fibroblasts", assay = "RNA", names.field = 1,
  names.delim = "_", meta.data = NULL)

```

QC on sample on its own

```{r}
Fibroblasts.ctrl[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.ctrl, pattern = "mt-")
Fibroblasts.dss[["percent.mt"]] <- PercentageFeatureSet(Fibroblasts.dss, pattern = "mt-")

#CTRL
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.ctrl <- PercentageFeatureSet(Fibroblasts.ctrl, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.ctrl, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.ctrl <- subset(Fibroblasts.ctrl, subset = nFeature_RNA > 300 & nFeature_RNA < 5000)


#DSS
# Visualize QC metrics as a violin plot
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

p# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
Fibroblasts.dss <- PercentageFeatureSet(Fibroblasts.dss, "^Rp", col.name = "percent_ribo")
VlnPlot(Fibroblasts.dss, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_ribo"), ncol = 4)

#Filtering from contaminated data
Fibroblasts.dss <- subset(Fibroblasts.dss, subset = nFeature_RNA > 300 & nFeature_RNA < 4000)
```


Downstream analysis. What Cell Types?
```{r}
# run sctransform
Fibroblasts.ctrl <- SCTransform(Fibroblasts.ctrl, vars.to.regress = "percent_ribo", verbose = FALSE, return.only.var.genes = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Fibroblasts.ctrl <- RunPCA(Fibroblasts.ctrl, verbose = FALSE)
Fibroblasts.ctrl <- RunUMAP(Fibroblasts.ctrl, dims = 1:20, verbose = FALSE)

Fibroblasts.ctrl <- FindNeighbors(Fibroblasts.ctrl, dims = 1:20, verbose = FALSE)
Fibroblasts.ctrl <- FindClusters(Fibroblasts.ctrl, verbose = FALSE, resolution = 0.5)
DimPlot(Fibroblasts.ctrl, label = TRUE)

# run sctransform
Fibroblasts.dss <- SCTransform(Fibroblasts.dss, vars.to.regress = "percent_ribo", verbose = FALSE,return.only.var.genes = FALSE)

# These are now standard steps in the Seurat workflow for visualization and clustering
Fibroblasts.dss <- RunPCA(Fibroblasts.dss, verbose = FALSE)
Fibroblasts.dss <- RunUMAP(Fibroblasts.dss, dims = 1:20, verbose = FALSE)

Fibroblasts.dss <- FindNeighbors(Fibroblasts.dss, dims = 1:20, verbose = FALSE)
Fibroblasts.dss <- FindClusters(Fibroblasts.dss, verbose = FALSE, resolution = 0.5)
DimPlot(Fibroblasts.dss, label = TRUE)
```

Annotate cells
```{r}
ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
ref.sub <- SCTransform(ref.sub)
unique(ref.sub$celltype_general)
ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")

raw_counts <- LayerData(Fibroblasts.ctrl, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Fibroblasts.ctrl <- AddMetaData(Fibroblasts.ctrl, pred$pruned.labels, col.name = 'celltype_integration')
DimPlot(Fibroblasts.ctrl, label = TRUE, group.by = "celltype_integration")
saveRDS(Fibroblasts.ctrl, "Fibroblasts_ctrl.rds")

#stroma.ref <-subset(Epi_Imm_Stro_colon, tissueLayer == "stroma")
#ref.sub<- JoinLayers(obj_Colonic_Stroma_DSS)
#ref.sub <- SCTransform(stroma.ref)
#unique(ref.sub$cellType)
#ref<- as.SingleCellExperiment(ref.sub, assay = "SCT")



raw_counts <- LayerData(Fibroblasts.dss, assay = "SCT") 

#Both ctrl and DSS in reference and a general celltype
pred <- SingleR(test = raw_counts, 
                  ref = ref, 
                  labels = ref$celltype_integration,
                  de.method = 'wilcox')
Fibroblasts.dss$celltype <- pred$labels[match(rownames(Fibroblasts.dss@meta.data), rownames(pred))]
Fibroblasts.dss <- AddMetaData(Fibroblasts.dss, pred$pruned.labels, col.name = 'celltype_integration')
DimPlot(Fibroblasts.dss, label = TRUE, group.by = "celltype_integration")
saveRDS(Fibroblasts.dss, "Fibroblasts_dss.rds")


```


###SMART-SEQ Data

#Colonic Cells. GSE148794

```{r}
library(readr)

df <- read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_cells/GSE148794_tc_ibd.count_table.tsv")

# Move the first column (named `...1`) to row names
df <- df %>%
  column_to_rownames(var = "...1")

metadata <-read_tsv("/Users/andson/Desktop/Mouse_Atlas_Datasets/Colon/WT+DSS/Colonic_cells/GSE148794_tc_ibd.metadata.tsv")

```

```{r}
# Split the 'Title' column into 'Treatment' and 'Sample' columns
metadata <- metadata %>%
  separate(Title, into = c("Prefix", "Treatment", "Sample_Suffix"), sep = "\\.", remove = FALSE) %>%
  separate(Sample_Suffix, into = c("Sample", "Rest"), sep = "_", remove = FALSE) %>%
  select(-Prefix, -Rest)


```


Marker information:
Clustering cells from healthy mice revealed 13 distinct clusters (Figure 3A and 3C ). We readily identified clusters showing specific expression of epithelial (Epcam and Krt19), pericyte (Rgs5 and Pdgfrb), vascular endothelial (Pecam1 / Cd31), lymphatic endothelial (Lyve1), and glial (S100b and Gfap) and hematopoietic cell markers (Cd52 and Ptprc / Cd45) (Figure 3C). Cluster 2, a small cluster of 32 cells, expressed markers associated with enteric smooth muscle (Myh11 and Des) and interstitial cells of Cajal (ICCs) (Kit and Ano1). Further examination of this cluster revealed its composition was two distinct sub-clusters consistent with ICCs and smooth muscle cells, respectively (Figure 3A). Other low-abundance clusters included enteric glial cells (14 cells) and pericytes (67 cells) (Figure 3A). The remaining 6 cell clusters (4–5, 10–13), comprising 3,391 cells or 89% of the dataset were fibroblast-like cells (FLCs) characterized by expression of the pan-fibroblast markers such as Dpt, Col6a2, and Col1a2 (Figure 3C). Clusters 4 and 5 also showed α-Sma expression, while only cluster 4 showed significant expression of smooth muscle myosin (Myh11) (Figure 3C). We readily identified these six populations as putative counterparts to the stromal cell populations in our human data by cluster marker expression (Table S6).



From:https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9399414/

PDGFRαhi fibroblasts: "Bmp3", "Bmp7" (Crypt top)


Pdgfra-NG2+Rgs5 + pericytes: "Rgs5"

Varying levels of Acta2, Myh11 and Des can help with the distinction between SMCs and myofibroblasts, but the two terms are sometimes used interchangeably in single cell RNA sequencing analyses (9–12, 14, 16, 18). Notably, the small intestine and colon display similar mesenchymal subsets with location-specific differences in their transcriptional profiles (5, 10).

SMC <- c("Acta2", "Cnn1", "Mylk")
SM <- c("Myh11", "Des")

Fibroblast_markers<- c("Acta2", "Dpt", "Col5a1", "Col1a2")

Sfrp1 and Grem1  markers of CD81+ fibroblasts. Cd81 (Crypt FB)
 PDGFRαloCD81- Fibroblasts
